### PR TITLE
Fix 25444 Fix possible nullpointer. Fix avoid begin() new transaction call in case of rollback.

### LINKED ISSUE #25444: Memory Leak caused by NullPointerException in ConnectorXAResource.getResourceHandle
I think a null check is missing to avoid this memory leak where TimerTask that check Transaction Timeout aren't removed. Result is an OutOfMemory occured during DB Connection Pool Problems.

The NPE seems to cause a infinite loop in timerthread, we see this stacktraces infinite in the server.log

see stacktrace of the NullPointer + screen of the Memory Leak 
```
[2025-04-03T19:18:46.404541+02:00] [GF 7.0.22] [INFO] [jdbc.exc_table_validation] [jakarta.enterprise.resource.resourceadapter.com.sun.gjc.spi] [tid: _ThreadID=3345 _ThreadName=uf-sdk-26167-3-rm-t-0] [levelValue: 800] [[
  RAR5074 : Table based validation detected invalid connection. Querying the table dual failed. Set resource-adapter log-level to FINE for exception stack trace.]]

[2025-04-03T19:18:46.404783+02:00] [GF 7.0.22] [SEVERE] [poolmgr.system_exception] [jakarta.enterprise.resource.resourceadapter.com.sun.enterprise.resource] [tid: _ThreadID=3345 _ThreadName=uf-sdk-26167-3-rm-t-0] [levelValue: 1000] [[
  RAR5031:System Exception
java.lang.NullPointerException: Cannot invoke "com.sun.enterprise.resource.ResourceHandle.getResourceState()" because "resourceHandle" is null
	at com.sun.enterprise.resource.ConnectorXAResource.getResourceHandle(ConnectorXAResource.java:213)
	at com.sun.enterprise.resource.ConnectorXAResource.rollback(ConnectorXAResource.java:179)
	at com.sun.enterprise.transaction.JavaEETransactionImpl.rollback(JavaEETransactionImpl.java:571)
	at com.sun.enterprise.resource.allocator.LocalTxConnectorAllocator.forceTransactionCompletion(LocalTxConnectorAllocator.java:140)
	at com.sun.enterprise.resource.allocator.LocalTxConnectorAllocator.destroyResource(LocalTxConnectorAllocator.java:110)
	at com.sun.enterprise.resource.pool.ConnectionPool.deleteResource(ConnectionPool.java:1054)
	at com.sun.enterprise.resource.pool.datastructure.RWLockDataStructure.removeAll(RWLockDataStructure.java:305)
	at com.sun.enterprise.resource.pool.ConnectionPool.emptyPool(ConnectionPool.java:1367)
	at com.sun.enterprise.resource.pool.ConnectionPool.doFailAllConnectionsProcessing(ConnectionPool.java:1245)
	at com.sun.enterprise.resource.pool.ConnectionPool.resourceErrorOccurred(ConnectionPool.java:1196)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.putbackResourceToPool(PoolManagerImpl.java:443)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.resourceErrorOccurred(PoolManagerImpl.java:393)
	at com.sun.enterprise.resource.listener.LocalTxConnectionEventListener.connectionErrorOccurred(LocalTxConnectionEventListener.java:101)
	at com.sun.gjc.spi.ManagedConnectionImpl.connectionErrorOccurred(ManagedConnectionImpl.java:785)
	at com.sun.gjc.spi.ManagedConnectionFactoryImpl.getInvalidConnections(ManagedConnectionFactoryImpl.java:300)
	at com.sun.enterprise.resource.allocator.AbstractConnectorAllocator.getInvalidConnections(AbstractConnectorAllocator.java:88)
	at com.sun.enterprise.resource.allocator.AbstractConnectorAllocator.isConnectionValid(AbstractConnectorAllocator.java:99)
	at com.sun.enterprise.resource.pool.ConnectionPool.isConnectionValid(ConnectionPool.java:750)
	at com.sun.enterprise.resource.pool.ConnectionPool.getResourceFromPool(ConnectionPool.java:823)
	at com.sun.enterprise.resource.pool.ConnectionPool.getUnenlistedResource(ConnectionPool.java:716)
	at com.sun.enterprise.resource.pool.ConnectionPool.internalGetResource(ConnectionPool.java:596)
	at com.sun.enterprise.resource.pool.ConnectionPool.getResource(ConnectionPool.java:465)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.getResourceFromPool(PoolManagerImpl.java:243)
	at com.sun.enterprise.resource.pool.PoolManagerImpl.getResource(PoolManagerImpl.java:154)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.getResource(ConnectionManagerImpl.java:355)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.internalGetConnection(ConnectionManagerImpl.java:298)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:183)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:157)
	at com.sun.enterprise.connectors.ConnectionManagerImpl.allocateConnection(ConnectionManagerImpl.java:151)
	at com.sun.gjc.spi.base.AbstractDataSource.getConnection(AbstractDataSource.java:99)
	at org.eclipse.persistence.sessions.JNDIConnector.connect(JNDIConnector.java:140)
	at org.eclipse.persistence.sessions.DatasourceLogin.connectToDatasource(DatasourceLogin.java:174)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.connectInternal(DatasourceAccessor.java:375)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.connectInternal(DatabaseAccessor.java:318)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.reconnect(DatasourceAccessor.java:618)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.reconnect(DatabaseAccessor.java:1820)
	at org.eclipse.persistence.internal.databaseaccess.DatasourceAccessor.incrementCallCount(DatasourceAccessor.java:346)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.basicExecuteCall(DatabaseAccessor.java:624)
	at org.eclipse.persistence.internal.databaseaccess.DatabaseAccessor.executeCall(DatabaseAccessor.java:569)
	at org.eclipse.persistence.internal.sessions.AbstractSession.basicExecuteCall(AbstractSession.java:2048)
	at org.eclipse.persistence.sessions.server.ClientSession.executeCall(ClientSession.java:311)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:280)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.executeCall(DatasourceCallQueryMechanism.java:266)
	at org.eclipse.persistence.internal.queries.DatasourceCallQueryMechanism.selectOneRow(DatasourceCallQueryMechanism.java:813)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRowFromTable(ExpressionQueryMechanism.java:2912)
	at org.eclipse.persistence.internal.queries.ExpressionQueryMechanism.selectOneRow(ExpressionQueryMechanism.java:2865)
	at org.eclipse.persistence.queries.ReadObjectQuery.executeObjectLevelReadQuery(ReadObjectQuery.java:563)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeDatabaseQuery(ObjectLevelReadQuery.java:1236)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:913)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
	at org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:448)
	at org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:3031)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1841)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1823)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1773)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.executeQuery(EntityManagerImpl.java:999)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.findInternal(EntityManagerImpl.java:941)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.find(EntityManagerImpl.java:817)
	at com.sun.enterprise.container.common.impl.EntityManagerWrapper.find(EntityManagerWrapper.java:387)
	at uep.game.betting.live.business.LiveManager.betstopAllEvents(LiveManager.java:556)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.glassfish.exousia.AuthorizationService.lambda$invokeBeanMethod$1(AuthorizationService.java:506)
	at org.glassfish.exousia.AuthorizationService.runInScope(AuthorizationService.java:586)
	at org.glassfish.exousia.AuthorizationService.invokeBeanMethod(AuthorizationService.java:506)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:297)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4193)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:604)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at org.jboss.weld.module.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:87)
	at org.jboss.weld.module.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:553)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:122)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:99)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:766)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:706)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:307)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4168)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4158)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:169)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:61)
	at jdk.proxy47/jdk.proxy47.$Proxy564.betstopAllEvents(Unknown Source)
	at uep.game.betting.live.business.__EJB31_Generated__LiveManager__Intf____Bean__.betstopAllEvents(Unknown Source)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.weld.util.reflection.Reflections.invokeAndUnwrap(Reflections.java:412)
	at org.jboss.weld.module.ejb.EnterpriseBeanProxyMethodHandler.invoke(EnterpriseBeanProxyMethodHandler.java:133)
	at org.jboss.weld.bean.proxy.EnterpriseTargetBeanInstance.invoke(EnterpriseTargetBeanInstance.java:54)
	at org.jboss.weld.module.ejb.InjectionPointPropagatingEnterpriseTargetBeanInstance.invoke(InjectionPointPropagatingEnterpriseTargetBeanInstance.java:69)
	at org.jboss.weld.bean.proxy.ProxyMethodHandler.invoke(ProxyMethodHandler.java:109)
	at uep.game.betting.live.business.LiveManager$Proxy$_$$_Weld$EnterpriseProxy$.betstopAllEvents(Unknown Source)
	at uep.game.betting.live.business.AbstractExternalFeedHandler.executeBetstop(AbstractExternalFeedHandler.java:341)
	at uep.game.betting.live.business.UnifiedFeedGlobalEventsListener.onProducerStatusChange(UnifiedFeedGlobalEventsListener.java:45)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.handleProducerStateChange(RecoveryManagerImpl.java:1059)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.flagProducerDown(RecoveryManagerImpl.java:956)
	at com.sportradar.unifiedodds.sdk.impl.recovery.RecoveryManagerImpl.onTimerElapsed(RecoveryManagerImpl.java:496)
	at com.sportradar.unifiedodds.sdk.impl.recovery.SingleRecoveryManagerSupervisor.lambda$scheduleSupervisionJob$0(SingleRecoveryManagerSupervisor.java:73)
	at com.sportradar.unifiedodds.sdk.impl.util.MdcScheduledExecutorService.lambda$setMdcContext$1(MdcScheduledExecutorService.java:499)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
]]
```
Image

### PR DESCRIPTION
**Pull request by escay**
- Fix #25444 
- Fix possible nullpointer. 
- Fix avoid begin new transaction call in case of rollback. 
- Add unit test to test variants of a rollback situation (transaction null vs nonXAResource null, and enlisted versus unenlisted).

### CODE REVIEW

File: appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java [hunk 1]
```diff
@@ -192,11 +192,16 @@ public boolean setTransactionTimeout(int seconds) throws XAException {
         return false;
     }
 
-    private ResourceHandle getResourceHandle() throws PoolingException {
+    // package default for unit test
```
Comment (pzygielo) [inline appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java:195]: I think this comment is unnecessary. If someone decides to restrict it back to private, the project (tests) should not compile.

File: appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+package com.sun.enterprise.resource.pool.mock;
+
+import jakarta.resource.ResourceException;
+import jakarta.resource.spi.LocalTransaction;
+
+/**
+ * Mock class without any implementation
+ */
+public class LocalTransactionMock implements LocalTransaction {
```
Comment (pzygielo) [inline appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java:25]: Just a question - can't `org.easymock:easymock`, which is dependency here already, be used instead all those handmade mocks?

File: appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java (hunk 1 above)
Comment (escay) [reply appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java:195]: I agree. It is a habit to document it: to try and prevent other classes in the same package to not use this method. If it is not documented, one might start using this method.

File: appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java (hunk 2 above)
Comment (escay) [reply appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java:25]: I have tried it. Similar to the one I created in ConnectionPoolTest: https://github.com/eclipse-ee4j/glassfish/blob/e02761182dcfdbc26c5892ea3ee46e563ccc404f/appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/ConnectionPoolTest.java#L100
But it added even more boilerplate and makes it hard for me to write and debug the unit test. All these connector-runtime classes require quite some layer on layer on layer logic that I cannot seem to be able to create a short test. 
For me using just java is the easiest approach in this module: generate 1 empty interface implementation mock class separate from the test with no logic in it, and add in the unit test an inner class where the required method is overridden.

File: appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java (hunk 1 above)
Comment (dmatej) [reply appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java:195]: Then it is better to push it to the javadoc ;-)

File: appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java (hunk 2 above)
Comment (dmatej) [reply appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/mock/LocalTransactionMock.java:25]: I am often struggling with those old "god-object-like" classes too, sometimes it is really easier to create simple class extending another one. There is a lot of work in front of us.
EasyMock is good, but when classes are too wired together and change behavior of each other it becomes a hell for mocking.
On the other hand as I see these classes are rather dummies, I would prefer EasyMock too in this case. Not sure, maybe I just don't see the problem, I would have to checkout and try. I believe you :-)

**Comment by dmatej** [comment]
The failed ssh test - please rebase to master, I already fixed that.

**Review by OndroMih — APPROVED** [review]
